### ISSUE TITLE
404 error caused by improper %-encoding

### ISSUE THREAD

**Issue by WaterFace** [gpodder/gpodder#861]
I'm trying to download the YKS podcast (rss https://feed.pippa.io/public/shows/yourkickstartersucks), but episodes prior to 105 fail to download with a 404 error.

For example, while trying to download episode 8, the exception it spits out in debug mode is 
`gpodder.download.gPodderDownloadHTTPError: ('//sphinx.acast.com/p/open/s/5d137ece8b774eb816199f63/e/tag:soundcloud,2010:tracks/314825464/media.mp3', 404, 'Not Found')`

However, if you paste the properly %-encoded url (https://sphinx.acast.com/p/open/s/5d137ece8b774eb816199f63/e/tag%3Asoundcloud%2C2010%3Atracks%2F314825464/media.mp3) into your browser, it works fine. For what it's worth, this is also the url inside the rss feed itself.

The reason that episodes 105 and later don't have this problem is that the url format changed (e.g. the url for episode 164 is https://sphinx.acast.com/p/open/s/5d137ece8b774eb816199f63/e/5f4d293775570c114b445fc7/media.mp3)

This behavior was observed on the most recent version from github cloned earlier today.

**Comment by auouymous** [comment]
The URL is properly escaped in the feed and gpodder downloads after decoding it. Most sites would decode the URL and match the escaped or unescaped versions. But acast.com matches before decoding.

@elelay Gpodder should use whatever URL string is in the feed and not do any decoding. But I don't know where in the download code it decodes the URL. The new requests code has the same issue.

**Comment by elelay** [comment]
Yes, this is caused by a step to encode [unicode characters in the url's path as ascii]()
What it does is decode the path the re-encode it, but with a list of 'safe' characters not needing to be escaped like '/' or ':' so the encoded '/' and ':' are not re-encoded.
All of this is in [util.py](https://github.com/gpodder/gpodder/blob/master/src/gpodder/util.py#L1973).
I still have to investigate if this is still required with requests.

**Comment by slikie** [comment]
I still have the same issue. Has there been any progress on this?

**Comment by Iizuki** [comment]
Just run into this with [Elucidations](https://feeds.acast.com/public/shows/57b498490b5f3f772a76004a) podcast, also on acast.

**Comment by elelay** [comment]
@Iizuki an episode in particular? I downloaded ep. 150 without issue

**Comment by Iizuki** [comment]
@elelay Huh?

1-85 fail for me every time:

![Elucidations download fail](https://github.com/user-attachments/assets/868c8899-4312-48ef-b98e-841fa7b8c4ff)


**Additional context:**

- Arch Linux
- gpodder 3.11.4
- python-mygpoclient 1.9-3
- python-podcastparser 0.6.10

**Comment by elelay** [comment]
confirmed for eg. episode 1.

**Comment by masoudd** [comment]
Episode 1 url in the [given feed](https://feeds.acast.com/public/shows/57b498490b5f3f772a76004a) is: 

`... url="https://sphinx.acast.com/p/open/s/57b498490b5f3f772a76004a/e/http%3A%2F%2Fhumstatic.uchicago.edu%2Fphilosophy%2Felucidations%2FCallard_DesireSatisfaction.mp3/media.mp3" ...`

And gpodder logs this error:

`[gpodder.download] ERROR: HTTP 404 while downloading "Episode 1: Agnes Callard discusses desire and satisfaction": 404 Client Error: Not Found for url: https://sphinx.acast.com/p/open/s/57b498490b5f3f772a76004a/e/http://humstatic.uchicago.edu/philosophy/elucidations/Callard_DesireSatisfaction.mp3/media.mp3`

So the problem is that gpodder turned `http%3A%2F%2F` to `http://`? Exactly inverse of issue #399. What is the right thing to do here? try the exact url given in the feed first and if it fails try the reencoded version?

**Comment by drakesobania** [comment]
Same problem. Also acast: https://feeds.acast.com/public/shows/5acd310d8f061bc45095260c

**Comment by drakesobania** [comment]
> Episode 1 url in the [given feed](https://feeds.acast.com/public/shows/57b498490b5f3f772a76004a) is:
> 
> `... url="https://sphinx.acast.com/p/open/s/57b498490b5f3f772a76004a/e/http%3A%2F%2Fhumstatic.uchicago.edu%2Fphilosophy%2Felucidations%2FCallard_DesireSatisfaction.mp3/media.mp3" ...`
> 
> And gpodder logs this error:
> 
> `[gpodder.download] ERROR: HTTP 404 while downloading "Episode 1: Agnes Callard discusses desire and satisfaction": 404 Client Error: Not Found for url: https://sphinx.acast.com/p/open/s/57b498490b5f3f772a76004a/e/http://humstatic.uchicago.edu/philosophy/elucidations/Callard_DesireSatisfaction.mp3/media.mp3`
> 
> So the problem is that gpodder turned `http%3A%2F%2F` to `http://`? Exactly inverse of issue #399. What is the right thing to do here? try the exact url given in the feed first and if it fails try the reencoded version?

I'm not having any issues with the URLs in AntennaPod. If I'm understanding this correctly (there's a solid chance I'm not), AntennaPod is doing exactly what you describe: https://github.com/AntennaPod/AntennaPod/blob/9acbafdba7ccbb6d74bb6a7335ff410f3f288a68/net/common/src/main/java/de/danoeh/antennapod/net/common/UriUtil.java

**Comment by Iizuki** [comment]
Python isn't quite my alley but apparently [this call](https://github.com/gpodder/gpodder/blob/d850e91ff50f4c78331fa057820984640982b83b/src/gpodder/download.py#L398) to `iri_to_url` is too much on these older Acast URLs?

Though it's unclear to me how the resolving a few lines prior works. 🤔

**Comment by masoudd** [comment]
> Python isn't quite my alley but apparently [this call](https://github.com/gpodder/gpodder/blob/d850e91ff50f4c78331fa057820984640982b83b/src/gpodder/download.py#L398) to `iri_to_url` is too much on these older Acast URLs?
> 
> Though it's unclear to me how the resolving a few lines prior works. 🤔

The resolving a few lines prior (`res = registry.download_url.resolve(config, None, episode, False)`) returns `None` for episode 1 of [given feed](https://feeds.acast.com/public/shows/57b498490b5f3f772a76004a).
It seems like the `registry` is used by files like `src/gpodder/youtube.py` or `src/gpodder/vimeo.py` to register custom downloaders that modify for example the url of a youtube video to an actual video file downloadable from youtube servers. None of those custom downloaders apply here. So the URL is modified in the `if url == episode.url:` branch and thus we go from `http%3A%2F%2F` to `http://` in the URL.

**Comment by Iizuki** [comment]
Thanks for the explanation!

So one approach could be to implement a custom downloader for Acast?

**Comment by masoudd** [comment]
> Thanks for the explanation!
> 
> So one approach could be to implement a custom downloader for Acast?

That would solve this specific Acast issue. But I think this should be solved in a more general way. Like trying the url unmodified first and if it fails try the modified one. Or the other way around.

**Comment by Iizuki** [comment]
Yeah that might be more productive in the long run. Though it seems like some amount of rewriting is in order?

**Comment by masoudd** [comment]
> Yeah that might be more productive in the long run. Though it seems like some amount of rewriting is in order?

Yes. Because the failure doesn't happen in the same function that is modifying the url. I think the url modification should be moved up to where the failure is detected and then it should try the unmodified url again
